### PR TITLE
Implement custom endpoints

### LINKED ISSUE #108: [provider] Implement endpoints
The `endpoints` parameter is not yet implemented. 

**_Please upvote this issue if you need this feature. This helps us to determine roadmap priorities._**

### PR DESCRIPTION
**Pull request by danielrbradley**
- Add a unit test running a real HTTP server to prove that requests are directed to the custom endpoint.
- Re-use test provider constructor from upgrade test work.
- Remove endpoints that we're not using and are irrelevant. These can be easily re-added in the future if needed.

Fixes #108

The credentials check was also used for populating the accountId which is used for invokes. If the accountId has not been fetched, fetch and cache it if the getAccountID is invoked.

### CODE REVIEW

File: provider/pkg/provider/provider.go [hunk 1]
```diff
@@ -111,6 +111,7 @@ type cfnProvider struct {
 	cctl *cloudcontrol.Client
 	ec2  *ec2.Client
```
Comment (mjeffryes) [inline provider/pkg/provider/provider.go:112]: looks like this is just used to implement a `getAZs` function. https://github.com/pulumi/pulumi-aws-native/blob/3585ccb59cfb21f6b8591764521c9652783599e1/provider/pkg/provider/provider.go#L511 I wonder if we should drop that in the future.

File: provider/pkg/provider/provider_endpoints_test.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+	})
+
+	t.Run("skipping credentials doesn't break getAccountId", func(t *testing.T) {
+		t.Parallel()
+		stsRequestCount := 0
+		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			stsRequestCount++
+			w.Write([]byte(stsGetCallerIdentityResponse))
+		}))
+		t.Cleanup(server.Close)
```
Comment (mjeffryes) [inline provider/pkg/provider/provider_endpoints_test.go:69]: nit: this feels like it should be using a mocking library or at least leverage some sort of helper method.

**Review by mjeffryes — APPROVED** [review]
Nifty! Feels pretty nice to be able to explicitly test the contract with these APIs in a few places. No blocking comments, though I think we'll want to be thoughtful before creating lots of tests using those server mocks to keep it maintainable.

File: provider/pkg/provider/provider.go (hunk 1 above)
Comment (danielrbradley) [reply provider/pkg/provider/provider.go:112]: Yeah, probably an easier way of doing this

File: provider/pkg/provider/provider_endpoints_test.go (hunk 2 above)
Comment (danielrbradley) [reply provider/pkg/provider/provider_endpoints_test.go:69]: Yeah, would be nice to use something a little more robust for testing in the future. Didn't want to just mock the http client but actually leverage the external http server for this specific endpoints case. Any other test for an E2E would probably go down the HTTP client mocking instead.

**Pull request merged**